### PR TITLE
Fix printing of templated records.

### PR DESCRIPTION
**Pull request by zahiraam**
Fixed the printing of templated argument list and added test case.

### CODE REVIEW

File: clang/lib/AST/TypePrinter.cpp [hunk 1]
```diff
@@ -2303,11 +2303,6 @@ printTo(raw_ostream &OS, ArrayRef<TA> Args, const PrintingPolicy &Policy,
     } else {
       if (!FirstArg)
         OS << Comma;
-      if (!Policy.SuppressTagKeyword &&
-          Argument.getKind() == TemplateArgument::Type &&
-          isa<TagType>(Argument.getAsType()))
-        OS << Argument.getAsType().getAsString();
-      else
```
Comment (zahiraam) [inline clang/lib/AST/TypePrinter.cpp:2310]: This code was wrongly added in https://github.com/intel/llvm/commit/6503b015d4a8 and was not reached by any tests. Some of our internal testing did hit this and realized the code was wrong. Created reproducers from internal tests to show that the code is wrong.

**Review by AaronBallman — APPROVED** [review]
Good catch, LGTM!

**Comment by zahiraam** [comment]
@AaronBallman Thanks!

**Comment by kazutakahirata** [comment]
I'm getting errors like:

```
llvm-project/clang/unittests/AST/DeclPrinterTest.cpp:1394:3: error: expression result unused [-Werror,-Wunused-value]
  [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
```

This file being a clang test, I am wondering if this is actually part of the test.  Shall we put `[[maybe_unused]]` here?

**Comment by zahiraam** [comment]
> I'm getting errors like:
> 
> ```
> llvm-project/clang/unittests/AST/DeclPrinterTest.cpp:1394:3: error: expression result unused [-Werror,-Wunused-value]
>   [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
> ```
> 
> This file being a clang test, I am wondering if this is actually part of the test. Shall we put `[[maybe_unused]]` here?

@kazutakahirata  When do you get this error?

**Comment by zahiraam** [comment]
Creating a PR with 
`[](PrintingPolicy [[maybe_unused]] & Policy) {
    Policy.SuppressTagKeyword = true;
  };`
  on those tests.

**Comment by kazutakahirata** [comment]
Looking at other examples, should the lambda expression be the last argument to `ASSERT_TRUE`?

**Comment by amy-kwan** [comment]
This also fails on PPC, as well: https://lab.llvm.org/buildbot/#/builders/36/builds/43998/steps/12/logs/stdio

**Comment by zahiraam** [comment]
oh! I think I know what the issue is. I have a fix for it. PR following up.

**Comment by mikaelholmen** [comment]
@zahiraam I still get
```
../../clang/unittests/AST/DeclPrinterTest.cpp:1394:3: error: expression result unused [-Werror,-Wunused-value]
  [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../clang/unittests/AST/DeclPrinterTest.cpp:1405:3: error: expression result unused [-Werror,-Wunused-value]
  [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../clang/unittests/AST/DeclPrinterTest.cpp:1416:3: error: expression result unused [-Werror,-Wunused-value]
  [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = true; };
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../clang/unittests/AST/DeclPrinterTest.cpp:1435:3: error: expression result unused [-Werror,-Wunused-value]
  [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../clang/unittests/AST/DeclPrinterTest.cpp:1455:3: error: expression result unused [-Werror,-Wunused-value]
  [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = true; };
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
5 errors generated.
```
This has been broken for quite a while now, can you please fix or revert?

**Comment by mikaelholmen** [comment]
> @zahiraam I still get
> 
> ```
> ../../clang/unittests/AST/DeclPrinterTest.cpp:1394:3: error: expression result unused [-Werror,-Wunused-value]
>   [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../../clang/unittests/AST/DeclPrinterTest.cpp:1405:3: error: expression result unused [-Werror,-Wunused-value]
>   [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../../clang/unittests/AST/DeclPrinterTest.cpp:1416:3: error: expression result unused [-Werror,-Wunused-value]
>   [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = true; };
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../../clang/unittests/AST/DeclPrinterTest.cpp:1435:3: error: expression result unused [-Werror,-Wunused-value]
>   [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = false; };
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../../clang/unittests/AST/DeclPrinterTest.cpp:1455:3: error: expression result unused [-Werror,-Wunused-value]
>   [](PrintingPolicy &Policy) { Policy.SuppressTagKeyword = true; };
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 5 errors generated.
> ```
> 
> This has been broken for quite a while now, can you please fix or revert?

Thank you @alinas for fixing the problem in 577e0ef94fb0 !

**Pull request merged**